hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/PartitionedMapTableResolver.java

```diff
@@ -174,11 +159,120 @@ private PartitionedMapTable createTable(
         }
     }
 
-    private static PartitionedMapTable emptyMap(String mapName) {
+    @SuppressWarnings("rawtypes")
+    private FieldsMetadata getHeapMapFields(MapServiceContext context, String name) {
+        for (PartitionContainer partitionContainer : context.getPartitionContainers()) {
+            // Resolve sample.
+            RecordStore<?> recordStore = partitionContainer.getExistingRecordStore(name);
+
+            if (recordStore == null) {
+                continue;
+            }
+
+            Iterator<Map.Entry<Data, Record>> recordStoreIterator = recordStore.iterator();
+
+            if (!recordStoreIterator.hasNext()) {
+                continue;
+            }
+
+            Map.Entry<Data, Record> entry = recordStoreIterator.next();
+
+            return getFieldMetadata(entry.getKey(), entry.getValue().getValue());
+        }
+
+        return FieldsMetadata.EMPTY_ERROR;
+    }
+
+    @SuppressWarnings("rawtypes")
+    private FieldsMetadata getHdMapFields(MapContainer mapContainer) {
+        if (!nodeEngine.getProperties().getBoolean(GLOBAL_HD_INDEX_ENABLED)) {
+            // Cannot resolve fields when concurrent indexes are disabled
+            return FieldsMetadata.HD_ERROR;
+        }
+
+        InternalIndex[] indexes = mapContainer.getIndexes().getIndexes();
+
+        if (indexes == null || indexes.length == 0) {
+            // Cannot resolve fields when the map doesn't have concurrent indexes
+            return FieldsMetadata.HD_ERROR;
+        }
+
+        InternalIndex index = indexes[0];
+
+        Iterator<QueryableEntry> entryIterator = index.getSqlRecordIterator();
+
+        if (!entryIterator.hasNext()) {
+            return FieldsMetadata.EMPTY_ERROR;
+        }
+
+        QueryableEntry entry = entryIterator.next();
+
+        return getFieldMetadata(entry.getKey(), entry.getValue());
+    }
+
+    private static PartitionedMapTable emptyError(String mapName) {
         QueryException error = QueryException.error(
             "Cannot resolve IMap schema because it doesn't have entries on the local member: " + mapName
         );
 
         return new PartitionedMapTable(mapName, error);
     }
+
+    private static PartitionedMapTable hdError(String mapName) {
+        QueryException error = QueryException.error("Cannot query the IMap \"" + mapName
```

Comment:
We already have a similar exception trying to run a map scan without index. Shall we reuse the code?

Reply:
The exception you mentioned is moved here

Reply:
Ahh, I see that the code has been moved.